File: src/main/java/com/github/dkoval/leetcode/problems/UniqueBinarySearchTrees.java
package com.github.dkoval.leetcode.problems;

/**
 * <a href="https://leetcode.com/problems/unique-binary-search-trees/">Unique Binary Search Trees</a>
 * <p>
 * Given an integer n, return the number of structurally unique BST's (binary search trees) which has exactly n nodes of unique values from 1 to n.
 * <p>
 * Constraints:
 * <p>
 * 1 <= n <= 19
 */
public interface UniqueBinarySearchTrees {

    int numTrees(int n);

    // O(N^2) time | O(N) space
    class UniqueBinarySearchTreesDPTopDown implements UniqueBinarySearchTrees {

        public int numTrees(int n) {
            // memo[i] is the number of unique BSTs with i nodes
            Integer[] memo = new Integer[n + 1];
            return numTrees(n, memo);
        }

        // input x is in [0 : N - 1] range
        // each input takes O(N) time -> overall O(N^2) time | O(N) space to allocate memo[]
        private int numTrees(int x, Integer[] memo) {
            if (x < 1) {
                return 1;
            }

            if (memo[x] != null) {
                return memo[x];
            }

            // l - number of nodes in the left subtree
            // r - number of nodes in the right subtree
            int count = 0;
            for (int l = 0; l < x; l++) {
                int r = x - l - 1;
                count += numTrees(l, memo) * numTrees(r, memo);
            }

            memo[x] = count;
            return count;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/UniqueBinarySearchTrees.kt
package com.github.dkoval.leetcode.challenge

/**
 * [Unique Binary Search Trees](https://leetcode.com/explore/featured/card/june-leetcoding-challenge/542/week-4-june-22nd-june-28th/3370/)
 *
 * Given n, how many structurally unique BST's (binary search trees) that store values 1 ... n?
 */
object UniqueBinarySearchTrees {

    // Resource: https://www.youtube.com/watch?v=4s7r3bO0hoU
    fun numTrees(n: Int): Int {
        if (n < 2) {
            return 1
        }

        // Let f'(x, y) is number of structurally unique BSTs with x nodes in the left and y nodes in the right subtrees respectively,
        // then the total number of structurally unique BSTs with n nodes is
        // f(n) = f'(0, n - 1) + f'(1, n - 2) + ... + f'(0, n - 1), where
        // f'(k, n - k - 1) = f(k) * f(n - k - 1)

        // dp[i] is the number of structurally unique BSTs with i nodes, i.e. f(i)
        val dp = IntArray(n + 1)
        dp[0] = 1
        dp[1] = 1
        for (i in 2..n) {
            for (k in 0 until i) {
                dp[i] += dp[k] * dp[i - k - 1]
            }
        }
        return dp.last()
    }
}